File: Gemini 3 Pro Thinking Solution.R
# Create an R script to generate a stacked bar plot of Global Deaths from 
# Particulate Matter Air Pollution (2023)

# Produced by GenAI with only minor edits by hand:
# - package loading modified to use pacman::p_load() and to add forcats & here
# - use of here() to improve flexibility in relative file path access
# - output file path stored as a variable to remove repetition path string
# - code line-wrapping modified to keep lines <= 80 characters long
# - plot annotation formatting modified to use a smaller font size

# This variation was created by Gemini 3 Pro Thinking from the following prompt:

# Goal: Create an R script to generate a stacked bar plot of Global Deaths from 
#       Particulate Matter Air Pollution (2023).
# 
# Data: Use the CSV file provided (possibly found within a "data" folder).
# 
# Plot Type: The plot should be a horizontal stacked bar plot x-axis is Deaths, 
#            y-axis is Cause of Death).
# 
# Data Filtering:
#   
#   Include only data where the Measure is "Percent of total deaths."
# 
# Only include death percentages >= 0.1 (after converting the raw value to a 
# percentage).
# 
# Aesthetics and Formatting:
#   
#   Stack the bars by the Risk Factor and use a pastel color palette.
# 
# Sort the bars from longest (top) to shortest (bottom).
# 
# Annotate each bar segment with its corresponding death percentage, rounded to 
# one decimal place.
# 
# Include a caption in the lower right margin stating the data source.
# 
# Output: Save the final plot as a PNG file in a newly created figures folder.
# 
# Note: Structure the script with clear steps and appropriate error handling.

# Load necessary libraries
if (!requireNamespace("pacman", quietly = TRUE)) install.packages("pacman")
pacman::p_load(tidyverse, forcats, here)

# 1. Setup and Data Loading
# Create 'figures' folder if it doesn't exist
if (!dir.exists("figures")) {
  dir.create("figures")
}

# Load the dataset
# Assuming the file is in a 'data' subfolder as specified
file_path <- here("data/IHME GBD Compare - Air Pollution Deaths.csv")

# Fallback: Check if file is in current directory if not found in 'data'
if (!file.exists(file_path)) {
  file_path <- here("IHME GBD Compare - Air Pollution Deaths.csv")
}

data <- read_csv(file_path, show_col_types = FALSE)

# 2. Data Filtering and Transformation
plot_data <- data %>%
  # Filter for the correct measure
  filter(Measure == "Percent of total deaths") %>%
  # Convert raw value to percentage
  mutate(Percent = Value * 100) %>%
  # Filter for significant percentages (>= 0.1%)
  filter(Percent >= 0.1)

# 3. Ordering Factors for Plotting
# We want to sort the bars from longest (top) to shortest (bottom).
# In ggplot, the y-axis fills from bottom to top. To put the longest bar at the 
# top, we need the factor level with the highest total percentage to be the 
# "last" level.
plot_data <- plot_data %>%
  group_by(`Cause of death or injury`) %>%
  mutate(Total_Percent = sum(Percent)) %>%
  ungroup() %>%
  mutate(`Cause of death or injury` = 
           fct_reorder(`Cause of death or injury`, Total_Percent))

# 4. Plotting
p <- ggplot(plot_data, aes(x = Percent, 
                           y = `Cause of death or injury`, 
                           fill = `Risk factor`)) +
  geom_col(width = 0.7) + # Stacked bar plot
  # Annotations: Percentage rounded to 1 decimal place
  geom_text(aes(label = sprintf("%.1f", Percent)), 
            position = position_stack(vjust = 0.5), 
            size = 2.5, 
            color = "black") +
  # Aesthetics
  scale_fill_brewer(palette = "Pastel1") + # Pastel color palette
  labs(title = "Global Deaths from Particulate Matter Air Pollution (2023)",
       x = "Deaths (% of Total)",
       y = "Cause of Death",
       caption = paste("Source: Institute for Health Metrics and Evaluation", 
                       "(IHME), GBD 2023")) +
  theme_minimal() +
  theme(
    plot.title = element_text(face = "bold", size = 14, hjust = 0.5),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.text = element_text(size = 10, color = "black"),
    panel.grid.major.y = element_blank() # Clean up y-axis grid lines
  )

# 5. Save Output
output_file <- here("figures/global_deaths_particulate_matter_2023.png")
ggsave(output_file, plot = p, width = 10, height = 6, dpi = 300)
message(paste("Script completed. Plot saved to", output_file, "."))

File: Gemini 3 Pro Thinking Solution.py
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import os

# To get these installed in RStudio (via reticulate), run in RStudio Terminal:
#   python -m ensurepip --upgrade
#   python -m pip install pandas
#   python -m pip install matplotlib

# Note: Gemini 3 Pro Thinking generated this python script as part of its 
# "thinking" process before generating the R script. I only edited the path
# to the data file, the filename of the PNG image, and the "Annotations" 
# section by reducing the font size and removing the "%" sign from the 
# annotations, to minimize overlap of the text. The rest was left as-is.

# Create figures directory
os.makedirs('figures', exist_ok=True)

# Load data
df = pd.read_csv('data/IHME GBD Compare - Air Pollution Deaths.csv')

# Filter
df_filtered = df[df['Measure'] == 'Percent of total deaths'].copy()
df_filtered['Percent'] = df_filtered['Value'] * 100

# Filter >= 0.1
df_filtered = df_filtered[df_filtered['Percent'] >= 0.1]

# Sort by Total Percent per Cause
total_per_cause = df_filtered.groupby('Cause of death or injury')['Percent'].sum().sort_values(ascending=True) # Ascending for horizontal bar plot (bottom to top)
# For top to bottom in plot, we usually plot 0 to N, so if we want longest at top, we might need to sort accordingly.
# Matplotlib plots from bottom up usually. So index 0 is bottom.
# If we want longest at top, the sort order should be ascending? Let's check.
# If I have [Small, Medium, Large], Small is at bottom. Large is at top.
# User wants longest (Large) at top. So Ascending sort of the index is correct.

ordered_causes = total_per_cause.index.tolist()
df_filtered['Cause of death or injury'] = pd.Categorical(df_filtered['Cause of death or injury'], categories=ordered_causes, ordered=True)

# Colors
# Pastel palette
# "Ambient particulate matter pollution", "Household air pollution from solid fuels"
risk_factors = df_filtered['Risk factor'].unique()
colors = ['#aec7e8', '#ffbb78', '#98df8a', '#ff9896'] # Pastel-ish colors
# Map risk factors to colors
color_map = {rf: c for rf, c in zip(risk_factors, colors)}

# Plotting
fig, ax = plt.subplots(figsize=(10, 6))

causes = ordered_causes
bottoms = np.zeros(len(causes))

for rf in risk_factors:
    # Extract data for this risk factor, ensuring alignment with 'causes'
    subset = df_filtered[df_filtered['Risk factor'] == rf]
    # Reindex to ensure we have all causes, fill missing with 0
    subset = subset.set_index('Cause of death or injury').reindex(causes).fillna({'Percent': 0}).reset_index()
    
    values = subset['Percent'].values
    
    bars = ax.barh(causes, values, left=bottoms, label=rf, color=color_map[rf], edgecolor='white')
    
    # Annotations
    for bar, val in zip(bars, values):
        if val > 0: # Only annotate if there is a bar
            width = bar.get_width()
            x_pos = bar.get_x() + width / 2
            y_pos = bar.get_y() + bar.get_height() / 2
            ax.text(x_pos, y_pos, f'{val:.1f}', ha='center', va='center', fontsize=7, color='black')

    bottoms += values

ax.set_xlabel('Percent of Total Deaths')
ax.set_ylabel('Cause of Death')
ax.set_title('Global Deaths from Particulate Matter Air Pollution (2023)')
ax.legend(title='Risk Factor')

# Caption
plt.figtext(0.99, 0.01, 'Source: IHME, Global Burden of Disease (2023)', horizontalalignment='right', fontsize=8)

plt.tight_layout()
plt.savefig('figures/global_deaths_particulate_matter_2023_python.png')
print("Plot saved.")

